test/distributed/fsdp/test_fsdp_flatten_params.py

```diff
@@ -68,152 +69,243 @@ def _get_shared_params_transformer(self, seed=0):
-        input = module.get_input(device, dtype)
-        return module(*input)
-
-    def _get_pnorm_after_step(self, module):
-        optim = torch.optim.SGD(module.parameters(), lr=0.01)
-        loss = self._get_output(module).sum()
-        loss.backward()
-        optim.step()
-        return torch.norm(torch.stack([p.detach().norm() for p in module.parameters()]))
-
-    def _test_num_params(self, module):
-        ref_num_params = sum(p.numel() for p in module.parameters())
-
-        params_to_flatten = list(module.parameters())
-        flat_module = FlattenParamsWrapper(
-            module,
-            params_to_flatten,
-            torch.device("cuda"),
-            self._get_default_config(),
-            False,
-        )
-        flat_num_params = sum(p.numel() for p in flat_module.parameters())
-
-        self.assertEqual(ref_num_params, flat_num_params)
-        self.assertEqual(flat_num_params, flat_module.flat_param.numel())
-
-    def _test_output(self, module):
-        ref_output = self._get_output(module)
-
-        params_to_flatten = list(module.parameters())
-        flat_module = FlattenParamsWrapper(
-            module,
-            params_to_flatten,
-            torch.device("cuda"),
-            self._get_default_config(),
-            False,
+    @skip_if_lt_x_gpu(1)
+    def test_partial_flattening(self):
+        """Tests flattening some submodules but not others."""
+        self.run_subtests(
+            {"half": [False, True]},
+            self._test_partial_flattening,
         )
-        flat_output = self._get_output(flat_module)
-        self.assertEqual(ref_output, flat_output)
 
-    def test_partial_flattening(self):
+    def _test_partial_flattening(self, half: bool):
         module = self._get_transformer()
-        num_params = sum(p.numel() for p in module.parameters())
-
-        params_to_flatten = list(module.encoder.layers[1].parameters()) + list(
-            module.decoder.layers[0].parameters()
+        if half:
+            module = module.half()
+        numel = sum(p.numel() for p in module.parameters())
+
+        encoder_1_params = list(module.encoder.layers[1].parameters())
+        decoder_0_params = list(module.decoder.layers[0].parameters())
+        params_to_flatten = encoder_1_params + decoder_0_params
+        num_params = [len(encoder_1_params), len(decoder_0_params)]
+        numel_to_flatten = sum(p.numel() for p in params_to_flatten)
+        module.encoder.layers[1] = FSDP(module.encoder.layers[1])
+        module.decoder.layers[0] = FSDP(module.decoder.layers[0])
+        flat_params = [
+            module.encoder.layers[1]._flat_param,
+            module.decoder.layers[0]._flat_param,
+        ]
+
+        self.assertEqual(sum(fp.numel() for fp in flat_params), numel_to_flatten)
+        self.assertEqual(sum(p.numel() for p in module.parameters()), numel)
+
+        # Check that flattened parameters have been replaced with a single
+        # `FlatParameter`
```

Comment:
These two are expected to be `1` because we register the `FlatParameter` to the wrapped module (i.e. `module.encoder.layers[1]` and `module.decoder.layers[0]`). This change happened in the 3/N PR in this stack.